pkg/reconcilermanager/controllers/reposync_controller.go

```diff
@@ -1139,60 +1146,108 @@ func copyConfigMapsToCms(ctx context.Context, cl client.Client, rs *v1beta1.Repo
+			continue
+		}
+		annotations := cm.GetAnnotations()
+		if value, found := annotations[repoSyncNameAnnotationKey]; !found || value != rs.Name {
+			// this ConfigMap was not created for this RepoSync
+			continue
+		}
+		if value, found := annotations[repoSyncNamespaceAnnotationKey]; !found || value != rs.Namespace {
+			// this ConfigMap was not created for this RepoSync
+			continue
+		}
+		if _, found := copiedConfigMapNames[cm.Name]; found {
+			// this ConfigMap is still needed for this RepoSync
+			continue
+		}
+		// none of the above conditions apply, so this ConfigMap is no longer needed, and we should delete it for
+		// cleanup
+		if err := cl.Delete(ctx, &cm); err != nil {
+			return nil, fmt.Errorf("failed to delete ConfigMap: %w", err)
+		}
 	}
 
 	return newValuesFileRefs, nil
 }
 
-func copyOneConfigMapToCms(ctx context.Context, cl client.Client, objRef types.NamespacedName) (string, error) {
+func copyOneConfigMapToCms(ctx context.Context, cl client.Client, objRef types.NamespacedName, rs *v1beta1.RepoSync) (string, error) {
 	var sourceConfigMap corev1.ConfigMap
 	err := cl.Get(ctx, objRef, &sourceConfigMap)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
 			// the source ConfigMap doesn't exist, so there is nothing to copy. The previous validation checks
 			// will already surface this as an error on the RepoSync, so we don't need to do anything here.
 			return "", nil
-		} else {
-			return "", fmt.Errorf("unexpected error fetching ConfigMap: %w", err)
 		}
+		return "", fmt.Errorf("unexpected error fetching ConfigMap: %w", err)
+
 	}
 
-	newNamespace := configsync.ControllerNamespace
-	newName := core.NsReconcilerPrefix + "-" + sourceConfigMap.Name + "-" + sourceConfigMap.Namespace
+	sourceConfigMapCopy := createSourceConfigMapCopy(rs, sourceConfigMap)
 
 	var existingConfigMapCopy corev1.ConfigMap
-	err = cl.Get(ctx, types.NamespacedName{Name: newName, Namespace: newNamespace}, &existingConfigMapCopy)
+	err = cl.Get(ctx, types.NamespacedName{Name: sourceConfigMapCopy.Name, Namespace: sourceConfigMapCopy.Namespace}, &existingConfigMapCopy)
+	if err == nil {
+		// the copied ConfigMap already exists, and is immutable, so we don't need to do anything.
+		return sourceConfigMapCopy.Name, nil
+	}
 	if err != nil && !apierrors.IsNotFound(err) {
 		return "", fmt.Errorf("unexpected error fetching configmap: %w", err)
 	}
-	if err == nil {
-		// the copied ConfigMap already exists, and is immutable, so we don't need to do anything.
-		return newName, nil
+
+	if err := cl.Create(ctx, &sourceConfigMapCopy); err != nil {
+		return "", fmt.Errorf("failed to create copy of configmap: %w", err)
 	}
+	return sourceConfigMapCopy.Name, nil
+}
+
+func createSourceConfigMapCopy(rs *v1beta1.RepoSync, sourceConfigMap corev1.ConfigMap) corev1.ConfigMap {
+	// use a hash to prevent the name from being too long
+	hash := sha1.Sum([]byte(fmt.Sprintf("%s:%s:%s", rs.Name, rs.Namespace, sourceConfigMap.Name)))
```

Comment:
This is the same [hash function that porch](https://github.com/GoogleContainerTools/kpt/blob/a20fed099c3cd20654125f9d95718f2a2f4db480/porch/pkg/git/package.go#L69) uses to create kubeobject names, but I'm happy to switch it to something else if config sync has a different standard

Reply:
According to [RFC-3174](https://www.rfc-editor.org/rfc/rfc3174.html), the hash is always 160 bits long. From my testing, this is a 40-character string in Go.